File: app/src/main/java/com/vovaplusexp/gpscontroller/utils/MathUtils.java
/*
 * GPS-Controller - Inertial Navigation with GPS Spoofing Detection
 * Copyright (C) 2024 VovaplusEXP
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 */
package com.vovaplusexp.gpscontroller.utils;

/**
 * Mathematical utility functions for inertial navigation
 */
public class MathUtils {
    
    /**
     * Project a point onto a line segment
     */
    public static double[] projectPointOnSegment(double px, double py, 
                                                  double x1, double y1, 
                                                  double x2, double y2) {
        double dx = x2 - x1;
        double dy = y2 - y1;
        
        if (dx == 0 && dy == 0) {
            return new double[]{x1, y1};
        }
        
        double t = ((px - x1) * dx + (py - y1) * dy) / (dx * dx + dy * dy);
        t = Math.max(0, Math.min(1, t));
        
        return new double[]{x1 + t * dx, y1 + t * dy};
    }
    
    /**
     * Calculate Haversine distance between two lat/lon points in meters
     */
    public static double haversineDistance(double lat1, double lon1, double lat2, double lon2) {
        double earthRadius = 6371000; // meters
        double dLat = Math.toRadians(lat2 - lat1);
        double dLon = Math.toRadians(lon2 - lon1);
        
        double a = Math.sin(dLat / 2) * Math.sin(dLat / 2) +
                   Math.cos(Math.toRadians(lat1)) * Math.cos(Math.toRadians(lat2)) *
                   Math.sin(dLon / 2) * Math.sin(dLon / 2);
        
        double c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
        
        return earthRadius * c;
    }
    
    /**
     * Rotate vector from device frame to world frame
     */
    public static float[] rotateToWorldFrame(float[] vector, float[] rotationMatrix) {
        float[] result = new float[3];
        
        result[0] = rotationMatrix[0] * vector[0] + rotationMatrix[1] * vector[1] + rotationMatrix[2] * vector[2];
        result[1] = rotationMatrix[3] * vector[0] + rotationMatrix[4] * vector[1] + rotationMatrix[5] * vector[2];
        result[2] = rotationMatrix[6] * vector[0] + rotationMatrix[7] * vector[1] + rotationMatrix[8] * vector[2];
        
        return result;
    }
    
    /**
     * Remove gravity from accelerometer reading
     */
    public static float[] removeGravity(float[] acceleration, float[] gravity) {
        float[] linear = new float[3];
        linear[0] = acceleration[0] - gravity[0];
        linear[1] = acceleration[1] - gravity[1];
        linear[2] = acceleration[2] - gravity[2];
        return linear;
    }
    
    /**
     * Apply low-pass filter for smoothing
     */
    public static float[] lowPassFilter(float[] input, float[] output, float alpha) {
        if (output == null) return input.clone();
        
        for (int i = 0; i < input.length; i++) {
            output[i] = output[i] + alpha * (input[i] - output[i]);
        }
        return output;
    }
    
    /**
     * Calculate vector magnitude
     */
    public static float magnitude(float[] vector) {
        float sum = 0;
        for (float v : vector) {
            sum += v * v;
        }
        return (float) Math.sqrt(sum);
    }
    
    /**
     * Normalize angle to [-180, 180] degrees
     */
    public static float normalizeAngle(float angle) {
        while (angle > 180) angle -= 360;
        while (angle < -180) angle += 360;
        return angle;
    }
    
    /**
     * Calculate angular difference
     */
    public static float angleDifference(float angle1, float angle2) {
        float diff = angle2 - angle1;
        return normalizeAngle(diff);
    }
}

File: app/src/main/java/com/vovaplusexp/gpscontroller/utils/MathUtils.kt
package com.vovaplusexp.gpscontroller.utils

import kotlin.math.*

fun lowPassFilter(input: FloatArray, output: FloatArray?, alpha: Float): FloatArray {
    val result = output ?: FloatArray(input.size)
    for (i in input.indices) {
        result[i] = result[i] + alpha * (input[i] - result[i])
    }
    return result
}

fun removeGravity(accel: FloatArray, gravity: FloatArray): FloatArray {
    val linearAcceleration = FloatArray(3)
    linearAcceleration[0] = accel[0] - gravity[0]
    linearAcceleration[1] = accel[1] - gravity[1]
    linearAcceleration[2] = accel[2] - gravity[2]
    return linearAcceleration
}

fun FloatArray.magnitude(): Float {
    return sqrt(this[0] * this[0] + this[1] * this[1] + this[2] * this[2])
}


object MathUtils {

    fun haversineDistance(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double {
        val r = 6371000 // Earth radius in meters
        val dLat = Math.toRadians(lat2 - lat1)
        val dLon = Math.toRadians(lon2 - lon1)
        val a = sin(dLat / 2) * sin(dLat / 2) +
                cos(Math.toRadians(lat1)) * cos(Math.toRadians(lat2)) *
                sin(dLon / 2) * sin(dLon / 2)
        val c = 2 * atan2(sqrt(a), sqrt(1 - a))
        return r * c
    }

    fun projectToSegment(px: Double, py: Double, ax: Double, ay: Double, bx: Double, by: Double): DoubleArray {
        val abx = bx - ax
        val aby = by - ay
        val apx = px - ax
        val apy = py - ay
        val ab2 = abx * abx + aby * aby
        val ap_ab = apx * abx + apy * aby
        var t = ap_ab / ab2
        t = max(0.0, min(1.0, t))
        return doubleArrayOf(ax + t * abx, ay + t * aby)
    }

    fun angleDifference(angle1: Float, angle2: Float): Float {
        var diff = (angle2 - angle1 + 180) % 360 - 180
        if (diff < -180) diff += 360
        return diff
    }

    fun rotateToWorldFrame(vector: FloatArray, rotationMatrix: FloatArray): FloatArray {
        val result = FloatArray(3)
        result[0] = rotationMatrix[0] * vector[0] + rotationMatrix[1] * vector[1] + rotationMatrix[2] * vector[2]
        result[1] = rotationMatrix[3] * vector[0] + rotationMatrix[4] * vector[1] + rotationMatrix[5] * vector[2]
        result[2] = rotationMatrix[6] * vector[0] + rotationMatrix[7] * vector[1] + rotationMatrix[8] * vector[2]
        return result
    }
}
